### ISSUE TITLE
Badge system implementation

### ISSUE THREAD

**Issue by Qiangong2** [Tuxemon/Tuxemon#869]
As battle areas/gyms are being added, I think it would be important to implement a badge system of some kind. This way we can limit the needed variables to access certain areas as well as contribute to the general story line

As of now, I see two ways of doing this:

1) Create items that are given to the player after every battle area victory
2) Have a single item (maybe a badge book?) that can be incremented after each win and update the description in the bag

------------------------------------------------------------------------------

Option 1 is probably the simplest of the two, just create different items and give them out after each win. However, I don't like this option. If we go with option 1, at every point we need to check how many badges someone has, we need to create a conditional statement for each badge. By the 10th badge this would mean at least 9 conditional statements in one event.

I prefer option 2 because we can just use a conditional statement such as `is badge:2` if you're at the third battle area, and so on. It will lessen the amount of repeat variables. However, for this we would need the ability to dynamically change descriptions of items in the bag.

--------------------------------------------------------------------------------

This is how I imagine option 2 to work:

A) you win battle
B) On first battle area, you are given badge book with first badge inside
C) Bag says something like:
```
This book contains the following badges:
Taba Town
```
D) At the second battle area, you are only given a badge. The variable assigned to the badge book is incremented by one, and the bag is updated to say the following:
```
This book contains the following badges:
Taba Town
City Park
```

-----------------------------------------------------------------

Regardless, for each option we'll need to have the ability to check whether someone has defeated an NPC

**Comment by jinna124** [comment]
I think this is a good idea, but can this be achieved be simply adding an item to the player's inventory and adding a badge each time you win, without making a lot of functions?

**Comment by bitcraft** [comment]
option #2 would require an additional IU to show the contents of some item (or a dynamic description), which makes me prefer option 1.  to implement, we could have a new condition which can return the number of badges in your inventory.  i don't really like the idea of wasting bag space for badges though, so the book idea does sound nice.  ill think on it but #2 is probably best long term

**Comment by xirsoi** [comment]
There could be multiple types of items (like in Gen2 and onward) that are in different compartments of your backpack. A badge section would hold just the badges.

I don't know why you'd need to have 9 different checks though @Qiangong2 .

**Comment by Qiangong2** [comment]
> There could be multiple types of items (like in Gen2 and onward) that are in different compartments of your backpack. A badge section would hold just the badges.
> 
> I don't know why you'd need to have 9 different checks though @Qiangong2 .

If we just required the last one, then someone could theoretically skip all the previous badges if they find a glitch. That's why you'd need 9 checks for the tenth battle

It would make it more tedious for debugging (as you'd need to go through everything), but I believe it's better to limit the potential issues.

**Comment by xirsoi** [comment]
Just gonna have to disagree with you there. If they find a glitch you fix the glitch, not make the rest of the game 10x more complicated.

**Comment by Qiangong2** [comment]
> Just gonna have to disagree with you there. If they find a glitch you fix the glitch, not make the rest of the game 10x more complicated.

I agree that we would need to fix the glitch, but it makes more sense to do both. That way, nobody can exploit the glitch before it's fixed.

**Comment by bitcraft** [comment]
> If we just required the last one, then someone could theoretically skip all the previous badges if they find a glitch. That's why you'd need 9 checks for the tenth battle

the game is an alpha state so we have to expect things will be glitched or broken at some point.  making 9 commands is a clear case of YAGNI, meaning "Ya Aint Gonna Need It" -- the effort to fix a theoretical problem is more than the negative effects of it actually happening, in theory.  better to stick to a good solid plan and fix issues as they arise....because a good implementation of a good design may render the problem obsolete.
 
 > There could be multiple types of items (like in Gen2 and onward) that are in different compartments of your backpack
 
 i didn't know this.  so if its a known thing, then i see no problem with #2 and we will wait until the UI can accommodate a "compartments" UI for badges.

**Comment by xirsoi** [comment]
Yeah, TMs and HMs (teach techniques) would be stored in there own compartment, for instance.
